pennylane/templates/layers/custom_entangler.py

```diff
@@ -0,0 +1,262 @@
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the ``BasicEntanglerLayers`` template.
+"""
+# pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
+from pennylane.templates.decorator import template
+from pennylane.ops import CNOT, RX
+from pennylane.templates import broadcast
+from pennylane.templates.utils import (
+    check_shape,
+    check_no_variable,
+    check_number_of_layers,
+    get_shape,
+    check_shapes,
+)
+from pennylane.wires import Wires
+
+
+@template
+def CustomEntanglerLayers(
+    rotation_weights, wires, rotation=None, coupling=None, coupling_weights=None, pattern=None
+):
+    r"""Layers consisting of one-parameter single-qubit rotations on each qubit, followed by a sequence of
+    parametrized double-qubit gates
+
+    The placement of double-qubit gates on the circuit is determined by a user-passed
+    `pattern`.
+
+    .. figure:: ../../_static/templates/layers/basic_entangler.png
+        :align: center
+        :width: 40%
+        :target: javascript:void(0);
+
+    The number of layers :math:`L` is determined by the first dimension of the
+    first element of the argument ``weights``.
+    When using a single wire, the template only applies the single
+    qubit gates in each layer.
+
+    .. note::
+
+        This template follows the convention of dropping the entanglement between the last and the first
+        qubit when using only two wires, so the entangler is not repeated on the same wires.
+        In this case, only one two-qubit gate is applied in each layer:
+
+        .. figure:: ../../_static/templates/layers/basic_entangler_2wires.png
+            :align: center
+            :width: 30%
+            :target: javascript:void(0);
+
+    Args:
+        rotation_weights (array[float]): array of weights with shape ``(L, len(wires))``, each weight is used as a parameter
+                                for the rotation
+        coupling_weights (array[float]): array of weights with shape ``(L, len(wires))``, each weight is used as a parameter
+                                for the coupling
+        wires (Iterable or Wires): Wires that the template acts on. Accepts an iterable of numbers or strings, or
+            a Wires object.
+        rotation (pennylane.ops.Operation): one-parameter single-qubit gate to use,
+                                            if ``None``, :class:`~pennylane.ops.RX` is used as default
+        pattern (?????): A keyword that determined how the double-qubit gates will be placed on the circuit.`pattern='ring'` is used
```

Comment:
Maybe could connect to the broadcast function here, so users know what the possible options are (and more maintainable if you link rather than reproduce the patterns here)